File: gajeong/Programmers/LV3/파괴되지않은건물.js
function solution(board, skill) {
  // N 가로 M 세로
  const N = board[0].length;
  const M = board.length;

  var answer = 0;
  const tmp = Array.from(Array(M + 1), () => Array(N + 1).fill(0));

  skill.forEach((el) => {
    let [mode, startY, startX, endY, endX, degree] = el;
    tmp[startY][startX] += mode === 1 ? -degree : degree;
    tmp[startY][endX + 1] += mode === 1 ? degree : -degree;
    tmp[endY + 1][startX] += mode === 1 ? degree : -degree;
    tmp[endY + 1][endX + 1] += mode === 1 ? -degree : degree;
  });

  // 누적합 구하기
  for (let y = 0; y < M; y++) {
    for (let x = 0; x < N; x++) {
      tmp[y][x + 1] += tmp[y][x];
    }
  }
  for (let x = 0; x < N; x++) {
    for (let y = 0; y < M; y++) {
      tmp[y + 1][x] += tmp[y][x];
    }
  }

  for (let y = 0; y < M; y++) {
    for (let x = 0; x < N; x++) {
      board[y][x] += tmp[y][x];
      if (board[y][x] > 0) answer += 1;
    }
  }

  return answer;
}

console.log(
  solution(
    [
      [5, 5, 5, 5, 5],
      [5, 5, 5, 5, 5],
      [5, 5, 5, 5, 5],
      [5, 5, 5, 5, 5],
    ],
    [
      [1, 0, 0, 3, 4, 4],
      [1, 2, 0, 2, 3, 2],
      [2, 1, 0, 3, 1, 2],
      [1, 0, 1, 3, 3, 1],
    ]
  )
);
console.log(
  solution(
    [
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ],
    [
      [1, 1, 1, 2, 2, 4],
      [1, 0, 0, 1, 1, 2],
      [2, 2, 0, 2, 0, 100],
    ]
  )
);

File: jiwoo/Programmers/파괴되지 않은 건물.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(vector<vector<int>> board, vector<vector<int>> skill)
{
    int answer = 0;

    int N, M;
    N = board.size();
    M = board[0].size();
    int b_copy[1002][1002];

    for (auto turn : skill)
    {
        int type = turn[0];
        int startx = turn[1];
        int starty = turn[2];
        int endx = turn[3];
        int endy = turn[4];
        int degree = turn[5];

        if (type == 1)
        {
            b_copy[startx][starty] -= degree;
            b_copy[endx + 1][endy + 1] -= degree;
        }
        else if (type == 2)
        {
            b_copy[startx][starty] += degree;
            b_copy[endx + 1][endy + 1] += degree;
        }
    }

    for (int i = 0; i <= N; i++)
    {
        for (int j = 0; j <= M; j++)
        {
            b_copy[i][j + 1] = b_copy[i][j] + b_copy[i][j + 1];
        }
    }
    for (int i = 0; i <= N; i++)
    {
        for (int j = 0; j <= M; j++)
        {
            b_copy[i + 1][j] = b_copy[i][j] + b_copy[i + 1][j];
        }
    }
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            board[i][j] += b_copy[i][j];
            if (board[i][j] > 0) answer++;
        }
    }

    return answer;
}

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}